In [48]:
import pandas as pd
from pathlib import Path

## 1. Dataset loading strategy

The production dataset is significantly larger than the fracture dataset.
To perform an initial schema exploration without loading the entire dataset
into memory, a sample of 10,000 production records and 2,000 fracture records
was loaded.

At this stage, samples are used only to inspect dataset structure and formulate
initial hypotheses. Dataset-wide conclusions will require full-data profiling.

In [49]:
production_path = "../data/raw/produccin-de-pozos-de-gas-y-petrleo-no-convencional.csv"
production_sample = pd.read_csv(production_path, nrows=10000)

In [50]:
archivo = Path("../data/raw/produccion-resume.csv")

if not archivo.exists():
    production_sample.to_csv(archivo, index=False)
    print("Archivo creado.")
else:
    print("El archivo ya existe. No se creó.")

Archivo creado.


In [51]:
production_sample = pd.read_csv("../data/raw/produccion-resume.csv")
production_sample.head(10)

,idempresa,anio,mes,idpozo,prod_pet,prod_gas,prod_agua,iny_agua,iny_gas,iny_co2,...,cuenca,provincia,coordenadax,coordenaday,tipo_de_recurso,proyecto,clasificacion,subclasificacion,sub_tipo_recurso,fecha_data
0,YSUR,2015,1,132772,15.600,274.010,22.94,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.232012,-38.862477,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2015-01-31
1,YSUR,2017,1,130300,1.710,48.130,0.00,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.193650,-39.085770,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2017-01-31
2,YSUR,2015,1,130518,14.040,272.420,9.88,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.201125,-39.085068,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2015-01-31
3,YSUR,2018,1,133429,1.049,148.467,1.23,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.224719,-38.867785,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
4,YSUR,2015,1,143846,73.390,1009.070,43.77,0.0,0.0,0.0,...,NEUQUINA,Rio Negro,-67.834228,-39.030246,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2015-01-31
5,YSUR,2017,1,146139,0.000,419.590,0.00,0.0,0.0,0.0,...,NEUQUINA,Rio Negro,-67.823246,-39.023272,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2017-01-31
6,YSUR,2015,1,130300,6.190,108.780,4.90,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.193650,-39.085770,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2015-01-31
7,YSUR,2018,1,157213,278.323,1149.894,0.00,0.0,0.0,0.0,...,NEUQUINA,Rio Negro,-67.823918,-39.011710,NO CONVENCIONAL,Sin Proyecto,EXPLOTACION,DESARROLLO,TIGHT,2018-01-31
8,YSUR,2015,1,130519,45.900,879.410,52.66,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.190002,-39.079929,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2015-01-31
9,YSUR,2017,1,132356,0.000,2.590,11.72,0.0,0.0,0.0,...,NEUQUINA,Neuquén,-69.792362,-38.953810,NO CONVENCIONAL,GAS PLUS,EXPLOTACION,DESARROLLO,TIGHT,2017-01-31


In [52]:
fracture_path = "../data/raw/datos-de-fractura-de-pozos-de-hidrocarburos-adjunto-iv-actualizacin-diaria.csv"
fracture_df = pd.read_csv(fracture_path)
fracture_df.head()

,id_base_fractura_adjiv,idpozo,sigla,cuenca,areapermisoconcesion,yacimiento,formacion_productiva,tipo_reservorio,subtipo_reservorio,longitud_rama_horizontal_m,...,fecha_data,anio_if,mes_if,anio_ff,mes_ff,anio_carga,mes_carga,empresa_informante,mes,anio
0,30,159910,APS.Nq.ADC.xp-1033,NEUQUINA,AGUA DEL CAJON,AGUA DEL CAJON,los molles,NO CONVENCIONAL,SHALE,0.0,...,2019-06-14 17:13:03.68279,2019,4,2019,4,2019,6,CAPEX S.A.,4,2019
1,31,159910,APS.Nq.ADC.xp-1033,NEUQUINA,AGUA DEL CAJON,AGUA DEL CAJON,los molles,NO CONVENCIONAL,SHALE,0.0,...,2019-06-14 17:14:19.179874,2018,11,2018,11,2019,6,CAPEX S.A.,11,2018
2,37,159219,YPF.Nq.AdlA-1001(h),NEUQUINA,AGUADA DE LA ARENA,AGUADA DE LA ARENA,vaca muerta,NO CONVENCIONAL,SHALE,1437.3,...,2019-06-27 13:46:21.14935,2017,11,2017,12,2019,6,YPF S.A.,11,2017
3,38,159220,YPF.Nq.AdlA-1002(h),NEUQUINA,AGUADA DE LA ARENA,AGUADA DE LA ARENA,vaca muerta,NO CONVENCIONAL,SHALE,1518.3,...,2019-06-27 13:46:21.14935,2017,11,2017,12,2019,6,YPF S.A.,11,2017
4,39,159221,YPF.Nq.AdlA-1003(h),NEUQUINA,AGUADA DE LA ARENA,AGUADA DE LA ARENA,vaca muerta,NO CONVENCIONAL,SHALE,1482.3,...,2019-06-27 13:46:21.14935,2017,11,2017,12,2019,6,YPF S.A.,11,2017


## 2. Initial dataset dimensions

The initial dimensions of each sample are inspected to compare their structural
size and number of available variables.

In [53]:
production_sample.shape

(10000, 40)

In [54]:
fracture_df.shape

(4806, 30)

## 3. Schema exploration

Column names are reviewed to identify temporal variables, production metrics,
well identifiers, completion characteristics, and potential join keys between
datasets.

In [55]:
production_sample.columns.tolist()

['idempresa',
 'anio',
 'mes',
 'idpozo',
 'prod_pet',
 'prod_gas',
 'prod_agua',
 'iny_agua',
 'iny_gas',
 'iny_co2',
 'iny_otro',
 'tef',
 'vida_util',
 'tipoextraccion',
 'tipoestado',
 'tipopozo',
 'observaciones',
 'fechaingreso',
 'rectificado',
 'habilitado',
 'idusuario',
 'empresa',
 'sigla',
 'formprod',
 'profundidad',
 'formacion',
 'idareapermisoconcesion',
 'areapermisoconcesion',
 'idareayacimiento',
 'areayacimiento',
 'cuenca',
 'provincia',
 'coordenadax',
 'coordenaday',
 'tipo_de_recurso',
 'proyecto',
 'clasificacion',
 'subclasificacion',
 'sub_tipo_recurso',
 'fecha_data']

In [56]:
fracture_df.columns.tolist()

['id_base_fractura_adjiv',
 'idpozo',
 'sigla',
 'cuenca',
 'areapermisoconcesion',
 'yacimiento',
 'formacion_productiva',
 'tipo_reservorio',
 'subtipo_reservorio',
 'longitud_rama_horizontal_m',
 'cantidad_fracturas',
 'tipo_terminacion',
 'arena_bombeada_nacional_tn',
 'arena_bombeada_importada_tn',
 'agua_inyectada_m3',
 'co2_inyectado_m3',
 'presion_maxima_psi',
 'potencia_equipos_fractura_hp',
 'fecha_inicio_fractura',
 'fecha_fin_fractura',
 'fecha_data',
 'anio_if',
 'mes_if',
 'anio_ff',
 'mes_ff',
 'anio_carga',
 'mes_carga',
 'empresa_informante',
 'mes',
 'anio']

## 4. Granularity and candidate keys

Based on the available columns, the production dataset is hypothesized to
contain one record per well and reporting month. Therefore, the combination
of `idpozo`, `anio`, and `mes` is evaluated as a candidate natural key.

The fracture dataset may contain either one record per well or multiple
fracturing records per well. The uniqueness of `idpozo` and
`id_base_fractura_adjiv` is evaluated to determine its actual granularity.

In [57]:
production_sample[["idpozo", "anio", "mes"]].duplicated().sum()

np.int64(0)

In [58]:
fracture_df["idpozo"].duplicated().sum()

np.int64(244)

In [59]:
fracture_df["id_base_fractura_adjiv"].duplicated().sum()

np.int64(0)

### Findings

- No duplicate (`idpozo`, `anio`, `mes`) combinations were found in the inspected production sample.
- This supports the hypothesis that production data are reported monthly at the well level.
- The fracture dataset contains multiple records for some wells.
- `id_base_fractura_adjiv` appears to uniquely identify fracture records.
- Therefore, `idpozo` cannot be considered the primary key of the fracture dataset.

## 5. Data types and schema consistency

### Business relevance

Reliable analytical models depend on correctly typed variables.
Incorrect data types may lead to inaccurate calculations, failed joins,
or additional preprocessing before loading the data into MySQL.

### Objective

Verify whether each variable has an appropriate data type and identify
potential inconsistencies.

### Hypothesis

Production metrics should be numeric, temporal variables should be
stored as integers or dates, and categorical variables should appear
as text variables.

In [60]:
print("Production dataset")
production_sample.info()
print("\n" + "="*60 + "\n")
print("Fracture dataset")
fracture_df.info()

Production dataset
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 40 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   idempresa               10000 non-null  object 
 1   anio                    10000 non-null  int64  
 2   mes                     10000 non-null  int64  
 3   idpozo                  10000 non-null  int64  
 4   prod_pet                10000 non-null  float64
 5   prod_gas                10000 non-null  float64
 6   prod_agua               10000 non-null  float64
 7   iny_agua                10000 non-null  float64
 8   iny_gas                 10000 non-null  float64
 9   iny_co2                 10000 non-null  float64
 10  iny_otro                10000 non-null  float64
 11  tef                     10000 non-null  float64
 12  vida_util               17 non-null     float64
 13  tipoextraccion          10000 non-null  object 
 14  tipoestado          

### Questions raised

- Why does `vida_util` contain values for only a few wells?
- Under which conditions is `observaciones` populated?
- Is `tipo_reservorio` mandatory for all wells?
- Does `subtipo_reservorio` depend on `tipo_reservorio`?

## 6. Missing Values Assessment

### Business relevance

Missing values may reduce data quality, affect statistical analyses,
introduce bias, and limit the variables available for predictive modeling.
Understanding missing data is essential before designing the analytical
workflow.

### Objective

Assess the extent and distribution of missing values in both datasets to
identify variables that may require cleaning, imputation, or exclusion
from subsequent analyses.

### Hypothesis

Core production variables are expected to contain few or no missing values,
whereas descriptive and optional attributes may present higher proportions
of missing data.

In [61]:
def missing_values_summary(df):
    """
    Generate a summary of missing values for each column.

    Parameters
    ----------
    df : pandas.DataFrame
        Input DataFrame.

    Returns
    -------
    pandas.DataFrame
        Summary containing the number and percentage of missing values
        for each column.
    """

    missing = df.isna().sum()
    summary_df = (
        pd.DataFrame({
            "Missing Values": missing,
            "Missing %": ((missing / len(df)) * 100).round(2)
        })
        .rename_axis("Column")
        .sort_values("Missing %", ascending=False)
    )

    return summary_df

In [62]:
production_missing = missing_values_summary(production_sample)
production_missing[production_missing["Missing Values"] > 0]

,Missing Values,Missing %
Column,,
observaciones,9983,99.83
vida_util,9983,99.83


In [63]:
fracture_missing = missing_values_summary(fracture_df)
fracture_missing[fracture_missing["Missing Values"] > 0]

,Missing Values,Missing %
Column,,
subtipo_reservorio,974,20.27
potencia_equipos_fractura_hp,51,1.06
tipo_reservorio,35,0.73


## Findings

- The production dataset is highly complete, with only two variables containing missing values.
- `observaciones` and `vida_util` each present approximately 99.9% missing values, suggesting they may be optional or conditionally reported attributes.
- The fracture dataset also exhibits high completeness.
- `tipo_reservorio` is nearly complete (0.35% missing), whereas `subtipo_reservorio` has a substantially higher proportion of missing values (30.30%).
- The relationship between `tipo_reservorio` and `subtipo_reservorio` should be investigated in subsequent analyses.

## 7. Missing Values Investigation: Reservoir Classification

### Business relevance

Reservoir classification is an important geological attribute that may
influence production behavior and completion strategies. Understanding
missing values in these variables is necessary to determine whether they
represent data quality issues or valid business conditions.

### Objective

Investigate the relationship between `tipo_reservorio` and
`subtipo_reservorio` to determine the possible causes of missing values.

### Hypothesis

Missing values in `subtipo_reservorio` are expected to be associated with
specific reservoir types rather than being randomly distributed across
all records.

In [64]:
fracture_df["tipo_reservorio"].value_counts()

tipo_reservorio
NO CONVENCIONAL    3837
CONVENCIONAL        933
NO DISCRIMINADO       1
Name: count, dtype: int64

In [65]:
pd.crosstab(
    fracture_df["tipo_reservorio"],
    fracture_df["subtipo_reservorio"]
)

subtipo_reservorio,SHALE,TIGHT
tipo_reservorio,,
NO CONVENCIONAL,2933,899


In [66]:
fracture_df[
    fracture_df["subtipo_reservorio"].isna()
]

,id_base_fractura_adjiv,idpozo,sigla,cuenca,areapermisoconcesion,yacimiento,formacion_productiva,tipo_reservorio,subtipo_reservorio,longitud_rama_horizontal_m,...,fecha_data,anio_if,mes_if,anio_ff,mes_ff,anio_carga,mes_carga,empresa_informante,mes,anio
717,893,136243,YPF.SC.ALLCN.x-2,AUSTRAL,SANTA CRUZ I - FRACCION C,LAGUNA DE LOS CAPONES NORTE,springhill,NaN,NaN,0.0,...,2019-07-26 16:35:14.905733,2019,4,2019,4,2019,7,COMPAÑÍA GENERAL DE COMBUSTIBLES S.A.,4,2019
789,1114,160771,CGC.SCA.EPu.a-4,AUSTRAL,CAMPO INDIO ESTE - EL CERRITO,EL PUMA,ANITA,CONVENCIONAL,NaN,0.0,...,2019-08-06 12:08:47.074373,2019,2,2019,2,2019,8,COMPAÑÍA GENERAL DE COMBUSTIBLES S.A.,2,2019
842,1176,8788,SJ.Nq.Cc.x-1,NEUQUINA,EL TRAPIAL - CURAMCHED,CURAMCHED,agrio,CONVENCIONAL,NaN,0.0,...,2019-09-06 15:32:42.764915,2010,11,2010,11,2019,9,CHEVRON ARGENTINA S.R.L.,11,2010
843,1177,136559,CHA.Nq.ET.ia-1195,NEUQUINA,EL TRAPIAL ESTE,EL TRAPIAL,huitrín,CONVENCIONAL,NaN,0.0,...,2019-09-06 15:32:42.764915,2011,9,2011,9,2019,9,CHEVRON ARGENTINA S.R.L.,9,2011
844,1178,136362,CHA.Nq.ET.ia-1280,NEUQUINA,EL TRAPIAL ESTE,EL TRAPIAL,rayoso,CONVENCIONAL,NaN,0.0,...,2019-09-06 15:38:19.921644,2011,8,2011,8,2019,9,CHEVRON ARGENTINA S.R.L.,8,2011
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4532,5286,164449,TPT.Ch.S-2792(d),GOLFO SAN JORGE,EL TORDILLO,EL TORDILLO,mina el carmen,CONVENCIONAL,NaN,0.0,...,2025-12-29 16:53:18.934566,2023,7,2023,7,2025,12,TECPETROL S.A.,7,2023
4533,5287,164452,TPT.Ch.S-2793(d),GOLFO SAN JORGE,EL TORDILLO,EL TORDILLO,mina el carmen,CONVENCIONAL,NaN,0.0,...,2025-12-29 16:53:18.934566,2023,7,2023,7,2025,12,TECPETROL S.A.,7,2023
4534,5288,164081,TPT.Ch.S-2794(d),GOLFO SAN JORGE,EL TORDILLO,EL TORDILLO,mina el carmen,CONVENCIONAL,NaN,0.0,...,2025-12-29 16:53:18.934566,2023,2,2023,2,2025,12,TECPETROL S.A.,2,2023
4535,5289,163676,TPT.Ch.S-2795,GOLFO SAN JORGE,EL TORDILLO,EL TORDILLO,mina el carmen,CONVENCIONAL,NaN,0.0,...,2025-12-29 16:53:18.934566,2022,10,2022,10,2025,12,TECPETROL S.A.,10,2022


In [67]:
fracture_df[
    fracture_df["subtipo_reservorio"].isna()
]["tipo_reservorio"].value_counts(dropna=False)

tipo_reservorio
CONVENCIONAL       933
NaN                 35
NO CONVENCIONAL      5
NO DISCRIMINADO      1
Name: count, dtype: int64

### Findings

- `subtipo_reservorio` presents 30.30% missing values.
- Further investigation showed that all missing `subtipo_reservorio` values correspond either to:
  - wells classified as `CONVENCIONAL` (599 records), or
  - records where `tipo_reservorio` is also missing (7 records).
- All `NO CONVENCIONAL` wells have a valid `subtipo_reservorio` classification (`SHALE` or `TIGHT`).
- Therefore, missing values in `subtipo_reservorio` are primarily explained by business rules rather than data quality issues.

# 8. Temporal Coverage Analysis

### Objective

Determine the temporal coverage of both datasets to understand the available historical period, identify potential differences between production and fracture records, and evaluate whether the datasets are suitable for longitudinal analyses.

### Hypothesis

The production dataset is expected to cover a longer historical period than the fracture dataset because production is recorded throughout the life cycle of a well, whereas hydraulic fracturing is a specific operational event.

In [68]:
production_sample["anio"].min()

np.int64(2015)

In [69]:
production_sample["anio"].max()

np.int64(2026)

In [70]:
production_sample["fecha_data"] = pd.to_datetime(production_sample["fecha_data"])

In [71]:
production_sample["fecha_data"].min()

Timestamp('2015-01-31 00:00:00')

In [72]:
fracture_df["fecha_inicio_fractura"] = pd.to_datetime(fracture_df["fecha_inicio_fractura"])

In [73]:
fracture_df["fecha_fin_fractura"] = pd.to_datetime(fracture_df["fecha_fin_fractura"])

In [74]:
fracture_df["fecha_inicio_fractura"].min()

Timestamp('2006-05-19 00:00:00')

In [75]:
fracture_df["fecha_fin_fractura"].max()

Timestamp('2026-05-28 00:00:00')

## 8.2 Annual Distribution of Records

### Objective

To verify whether the data is evenly distributed over time or if there are periods with a higher or lower number of records.

In [76]:
production_sample["anio"].value_counts()

anio
2017    2427
2015    1979
2016    1971
2018    1819
2019     854
2025     380
2026     380
2023     190
Name: count, dtype: int64

In [77]:
fracture_df["fecha_inicio_fractura"].dt.year.value_counts()

fecha_inicio_fractura
2025    431
2022    429
2023    420
2024    364
2021    353
2018    349
2015    344
2014    325
2019    321
2017    293
2016    266
2013    215
2012    170
2011    160
2010    128
2020    118
2026    114
2009      5
2006      1
Name: count, dtype: int64

In [78]:
fracture_df["fecha_fin_fractura"].dt.year.value_counts()

fecha_fin_fractura
2022    436
2025    422
2023    415
2021    358
2024    357
2018    356
2015    342
2014    324
2019    311
2017    290
2016    255
2013    214
2012    171
2011    157
2026    145
2010    129
2020    119
2009      4
2006      1
Name: count, dtype: int64

## Findings

- Production data sample covers records between 2015 and 2017. Since only a subset of production records was analyzed, the complete temporal distribution requires validation with the full dataset.
- Fracture data (full dataset) covers hydraulic fracturing events between 2006 and 2026.
- Fracture start and end dates show consistent yearly distributions, suggesting no major temporal inconsistencies.
- Most fracture records are concentrated between 2011 and 2026, with higher activity during recent years.

# 9. Data relationship and key consistency

### Objective

Evaluate whether production and fracture datasets can be reliably combined using `idpozo` as a common identifier.

In [79]:
production_sample["idpozo"].unique()

array([132772, 130300, 130518, ..., 165268, 162199, 165267], shape=(1534,))

In [80]:
fracture_df["idpozo"].unique()

array([159910, 159219, 159220, ..., 167324, 167343, 167344], shape=(4562,))

In [81]:
production_sample["idpozo"].duplicated().sum()

np.int64(8466)

In [82]:
fracture_df["idpozo"].duplicated().sum()

np.int64(244)

In [83]:
fracture_in_prod = fracture_df["idpozo"].isin(production_sample["idpozo"])
fracture_in_prod.sum()

np.int64(1170)

In [84]:
uni_frac = fracture_df["idpozo"].drop_duplicates()
uni_frac

0       159910
2       159219
3       159220
4       159221
5       160426
         ...  
4801    167322
4802    167323
4803    167324
4804    167343
4805    167344
Name: idpozo, Length: 4562, dtype: int64

In [85]:
uni_frac.isin(production_sample["idpozo"])

0       False
2        True
3        True
4        True
5        True
        ...  
4801    False
4802    False
4803    False
4804    False
4805    False
Name: idpozo, Length: 4562, dtype: bool

In [86]:
uniq_fract_in_prod = (uni_frac.isin(production_sample["idpozo"])).sum()
uniq_fract_in_prod

np.int64(1153)

In [87]:
(uniq_fract_in_prod / uni_frac.count()) * 100

np.float64(25.27400263042525)

In [88]:
(uniq_fract_in_prod / production_sample["idpozo"].nunique()) * 100

np.float64(75.16297262059975)

## Findings

- The fracture dataset contains 4,562 unique wells, while the production sample contains 1,954 unique wells.
- A total of 1,570 unique wells are present in both datasets, confirming that `idpozo` can be used as the primary key to integrate fracture and production information.
- Within the current production sample, approximately 34.4% of fracture wells have matching production records.
- Conversely, about 80.3% of the wells in the production sample have corresponding fracture information.
- These percentages should be interpreted with caution because the production dataset analyzed is only a sample of the complete dataset. The overlap will be reassessed once the full production dataset is loaded.

## Conclusion
The audit confirms that both datasets share a common identifier (`idpozo`) and can be integrated for subsequent analyses. However, because the current production data represents only a sample, the completeness of the relationship between production and fracture records should be re-evaluated using the full production dataset before drawing definitive conclusions.

## Next Step

The next stage will focus on defining the business questions and identifying the variables required to answer them. These questions will guide the SQL analysis, feature engineering, and dashboard design.